build/gen_stub.php

```diff
@@ -46,28 +46,54 @@ function processDirectory(string $dir, Context $context): array {
     return $fileInfos;
 }
 
-function processStubFile(string $stubFile, Context $context): ?FileInfo {
+function processStubFile(string $stubFile, Context $context, bool $includeOnly = false): ?FileInfo {
     try {
         if (!file_exists($stubFile)) {
             throw new Exception("File $stubFile does not exist");
         }
 
-        $stubFilenameWithoutExtension = str_replace(".stub.php", "", $stubFile);
-        $arginfoFile = "{$stubFilenameWithoutExtension}_arginfo.h";
-        $legacyFile = "{$stubFilenameWithoutExtension}_legacy_arginfo.h";
+        if (!$includeOnly) {
+            $stubFilenameWithoutExtension = str_replace(".stub.php", "", $stubFile);
+            $arginfoFile = "{$stubFilenameWithoutExtension}_arginfo.h";
+            $legacyFile = "{$stubFilenameWithoutExtension}_legacy_arginfo.h";
 
-        $stubCode = file_get_contents($stubFile);
-        $stubHash = computeStubHash($stubCode);
-        $oldStubHash = extractStubHash($arginfoFile);
-        if ($stubHash === $oldStubHash && !$context->forceParse) {
-            /* Stub file did not change, do not regenerate. */
-            return null;
+            $stubCode = file_get_contents($stubFile);
+            $stubHash = computeStubHash($stubCode);
+            $oldStubHash = extractStubHash($arginfoFile);
+            if ($stubHash === $oldStubHash && !$context->forceParse) {
+                /* Stub file did not change, do not regenerate. */
+                return null;
+            }
         }
 
-        initPhpParser();
-        $fileInfo = parseStubFile($stubCode);
-        $constInfos = $fileInfo->getAllConstInfos();
-        $context->allConstInfos = array_merge($context->allConstInfos, $constInfos);
+        if (!$fileInfo = $context->parsedFiles[$stubFile] ?? null) {
+            initPhpParser();
+            $fileInfo = parseStubFile($stubCode ?? file_get_contents($stubFile));
+            $context->parsedFiles[$stubFile] = $fileInfo;
+
+            foreach ($fileInfo->dependencies as $dependency) {
+                // TODO add header search path for extensions?
+                $prefixes = [dirname($stubFile) . "/", ""];
```

Comment:
Hmm, how does it work? The `Zend/zend_attributes.stub.php` is two directories deeper than `ext/zend_test/test.stub.php`. Wouldn't is be possible to support relative paths *only*, like `require "../../Zend/zend_attributes.stub.php";`?

Reply:
I was imagining something like `include_path=,;/path/to/source/root/`. At least once we add php stubs to headers, we definitely need to include such a "magic" location, whereby the source can be addressed.

I.e. in an external ext, you don't have any proper relative path to use. And the design goal here was that you can still do an in-tree build of exts without any changes. If we were to remove this in php-src, everytime you want to do in-tree development of an ext. you'd have to prepend `../../` to any include, which is not great.